xmipp

```diff
@@ -210,70 +228,102 @@ def cleanBinaries():
     runJob('find . -iname "*.pyc" -delete')
     runJob("rm -rf xmipp.conf build")
 
-def cloneOrCheckout(repo, branch, doPull, ok):
-    repoDir = "src/%s"%repo
-    print("...on %s" % repoDir)
+def checkout(branch, doPull):
+    log = []
+    words = ['working directory clean', 'working tree clean']
+    runJob('git status', show_output=False, show_command=False, log=log)
+    if not any(w in l for w in words for l in log):
+        print(red('working branch detected: aborting git checkout and git pull'))
+        return False
+    else:
+        r, currentBranch = getCurrentBranch()
+        if currentBranch != branch:
+            if not runJob("git checkout %s" % branch):
+                print(red("Cannot checkout branch '%s'. Remaining on the branch '%s'."
+                      % (branch, currentBranch)))
+                return False
```

Comment:
It can continue (return True). For instance, if you have a working directory in xmipp and xmippCore repo but not in the xmippViz, then you want to continue. Just a warning it's enough (the red message).

Reply:
I disagree.  
If it manages to checkout only one out of two branches, than the command did not finish successfully.

Reply:
But most of the times, we have working branches in only one or two of the repositories. For instance, I'm working in a new protocol. Thus, I have a working branch in xmipp, xmippCore and scipion-em-xmipp called 'dm_newProt'. The devel branch is fine for the xmippViz repo. So, the `./xmipp all br=dm_newProt` command should work for me.

Of course, if the current branch is not devel but it's an strange one, this causes a problem. For that reason the warning in red.

Reply:
The same for the case "Working directory found...".

Reply:
Well, we can discuss this... We can assume that if you are developing, you should work with ./xmipp git checkout whatever and when all is as you want, then run ./xmipp all N=# without the 'br' flag.

Reply:
I am not convinced.
Should we return True, it might lead to very hard-to-find issues once we call xmipp from some batch.
Imagine you want to change your working branches and compile. Xmipp script would do that for you (returns 0 for checkout), and while compiling your will get strange compilation errors, because one of the repositories has not been updated properly due to some forgotten files.
On the other hand, should the script return e.g. -1 (some error code), you can handle it properly (stash changes, force checkout etc.).